{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "leading-merchandise",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import dask.dataframe as dd\n",
    "from pathlib import Path\n",
    "import subprocess as sp\n",
    "import math\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import numpy as np\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "chubby-arcade",
   "metadata": {},
   "outputs": [],
   "source": [
    "def genome_TSS_annotation(genome_path_gtf, TSS_gtf):\n",
    "    \"\"\"\n",
    "    Takes a ensemble genome gtf file as input, and returns a smaller gtf file containing only the genes and\n",
    "    their TSS locations\n",
    "    \"\"\"\n",
    "    Path(f\"{TSS_gtf}\").touch()\n",
    "    tempfile = os.getcwd() + \"/temp_TSS_genome_annotation.txt\"\n",
    "    Path(tempfile).touch()\n",
    "    os.remove(f\"{TSS_gtf}\")\n",
    "    # i = 0\n",
    "    with open(f\"{tempfile}\", \"a\") as output_gtf:\n",
    "        with open(genome_path_gtf) as f:\n",
    "            for _ in range(5):\n",
    "                next(f)\n",
    "            for line in f:\n",
    "                line = line.strip()\n",
    "                line = line.split()\n",
    "                gene_name = line[13]\n",
    "                coding_stuff = line[17]\n",
    "                strand_info = line[6]\n",
    "                #             if i == 1000:\n",
    "                #                 break\n",
    "                #             i = i + 1\n",
    "                gene_name = gene_name.rstrip(\";\")\n",
    "                coding_stuff = coding_stuff.rstrip(\";\")\n",
    "                # coding_stuff = coding_stuff.rstrip(\";\")\n",
    "                if str(coding_stuff) == '\"protein_coding\"':\n",
    "                    if line[6] == \"+\":\n",
    "                        start = line[3]\n",
    "                        end = int(start) + 1\n",
    "                        output_gtf.write(\n",
    "                            line[0]\n",
    "                            + \"\\t\"\n",
    "                            + str(start)\n",
    "                            + \"\\t\"\n",
    "                            + str(end)\n",
    "                            + \"\\t\"\n",
    "                            + gene_name\n",
    "                            + \"\\n\"\n",
    "                        )\n",
    "                    else:\n",
    "                        start = int(line[4]) - 1\n",
    "                        end = line[4]\n",
    "                        output_gtf.write(\n",
    "                            line[0]\n",
    "                            + \"\\t\"\n",
    "                            + str(start)\n",
    "                            + \"\\t\"\n",
    "                            + str(end)\n",
    "                            + \"\\t\"\n",
    "                            + gene_name\n",
    "                            + \"\\n\"\n",
    "                        )\n",
    "    sp.check_call(\n",
    "        f\"nice -5 bedtools sort \" f\"-i {tempfile} \" f\"> {TSS_gtf}\", shell=True,\n",
    "    )\n",
    "    os.remove(tempfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "local-memphis",
   "metadata": {},
   "outputs": [],
   "source": [
    "def TSS_window_to_region(genome_TSS_gtf, bedregion_file, TSS_window):\n",
    "    \"\"\"\n",
    "    Takes a genome TSS gtf file or bedfile as input (see the 'genome_TSS_GTF' script) together with a bed file \n",
    "    with regions and a window length.\n",
    "    It wil subsequently take a window around the TSS, and list all regions overlapping with this window. It will\n",
    "    furhtermore calculate the distance of te bedregion to the TSS.\n",
    "    It will return a dataframe containing all these results.\n",
    "    \"\"\"\n",
    "    sp.check_call(\n",
    "        f\"nice -10 bedtools window \"\n",
    "        f\"-w {TSS_window} \"\n",
    "        f\"-a {genome_TSS_gtf} \"\n",
    "        f\"-b {bedregion_file} \"\n",
    "        f\"> tmp_bedtool_window.txt\",\n",
    "        shell=True,\n",
    "    )\n",
    "    pd_table = pd.read_table(\n",
    "        \"tmp_bedtool_window.txt\",\n",
    "        index_col=None,\n",
    "        names=[\n",
    "            \"chr_TSS\",\n",
    "            \"start_TSS\",\n",
    "            \"end_TSS\",\n",
    "            \"gene_name\",\n",
    "            \"Chrom\",\n",
    "            \"ChromStart\",\n",
    "            \"ChromEnd\",\n",
    "        ],\n",
    "        dtype={\n",
    "            \"chr_TSS\": str,\n",
    "            \"start_TSS\": int,\n",
    "            \"end_TSS\": int,\n",
    "            \"gene_name\": str,\n",
    "            \"Chrom\": str,\n",
    "            \"ChromStart\": int,\n",
    "            \"ChromEnd\": int,\n",
    "        },\n",
    "    )\n",
    "    os.remove(\"tmp_bedtool_window.txt\")\n",
    "    distance = (\n",
    "        []\n",
    "    )  # calculate the distance based on the TSS start end region start site, check which one is upstream for distance calc.\n",
    "    for x, y in zip(pd_table[\"start_TSS\"], pd_table[\"ChromStart\"]):\n",
    "        if x > y:\n",
    "            distance.append(x - y)\n",
    "        elif y > x:\n",
    "            distance.append(y - x)\n",
    "        else:\n",
    "            distance.append(\"NaN\")\n",
    "    pd_table[\"distance\"] = distance\n",
    "    return pd_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "continental-clear",
   "metadata": {},
   "outputs": [],
   "source": [
    "def distance_weight(include_promoter=False, include_enhancer=True, alpha=1e5, padding=100000, keep1=5000, remove=2000):\n",
    "    \"\"\"\n",
    "    Built weight distribution from TSS. (From Ananse)\n",
    "    \"\"\"\n",
    "    # alpha is half site, default setting is 1e5, which means at 1e5 position weight is 0.5\n",
    "    # padding is the full range we used\n",
    "    # remove is promoter removed range\n",
    "    # keep1 is keep full binding score range\n",
    "    u = -math.log(1.0 / 3.0) * 1e5 / alpha\n",
    "    if include_promoter and include_enhancer:\n",
    "        weight1 = pd.DataFrame(\n",
    "            {\"weight\": [1 for z in range(1, remove + 1)], \"dist\": range(1, remove + 1)}\n",
    "        )\n",
    "        weight2 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [1 for z in range(remove + 1, keep1 + 1)],\n",
    "                \"dist\": range(remove + 1, keep1 + 1),\n",
    "            }\n",
    "        )\n",
    "        weight3 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [\n",
    "                    2.0\n",
    "                    * math.exp(-u * math.fabs(z) / 1e5)\n",
    "                    / (1.0 + math.exp(-u * math.fabs(z) / 1e5))\n",
    "                    for z in range(1, padding - keep1 + 1)\n",
    "                ],\n",
    "                \"dist\": range(keep1 + 1, padding + 1),\n",
    "            }\n",
    "        )\n",
    "    elif not include_promoter and include_enhancer:\n",
    "        weight1 = pd.DataFrame(\n",
    "            {\"weight\": [0 for z in range(1, remove + 1)], \"dist\": range(1, remove + 1)}\n",
    "        )\n",
    "        weight2 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [1 for z in range(remove + 1, keep1 + 1)],\n",
    "                \"dist\": range(remove + 1, keep1 + 1),\n",
    "            }\n",
    "        )\n",
    "        weight3 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [\n",
    "                    2.0\n",
    "                    * math.exp(-u * math.fabs(z) / 1e5)\n",
    "                    / (1.0 + math.exp(-u * math.fabs(z) / 1e5))\n",
    "                    for z in range(1, padding - keep1 + 1)\n",
    "                ],\n",
    "                \"dist\": range(keep1 + 1, padding + 1),\n",
    "            }\n",
    "        )\n",
    "    elif include_promoter and not include_enhancer:\n",
    "        weight1 = pd.DataFrame(\n",
    "            {\"weight\": [1 for z in range(1, remove + 1)], \"dist\": range(1, remove + 1)}\n",
    "        )\n",
    "        weight2 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [0 for z in range(remove + 1, keep1 + 1)],\n",
    "                \"dist\": range(remove + 1, keep1 + 1),\n",
    "            }\n",
    "        )\n",
    "        weight3 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [0 for z in range(1, padding - keep1 + 1)],\n",
    "                \"dist\": range(keep1 + 1, padding + 1),\n",
    "            }\n",
    "        )\n",
    "    else:\n",
    "        weight1 = pd.DataFrame(\n",
    "            {\"weight\": [0 for z in range(1, remove + 1)], \"dist\": range(1, remove + 1)}\n",
    "        )\n",
    "        weight2 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [0 for z in range(remove + 1, keep1 + 1)],\n",
    "                \"dist\": range(remove + 1, keep1 + 1),\n",
    "            }\n",
    "        )\n",
    "        weight3 = pd.DataFrame(\n",
    "            {\n",
    "                \"weight\": [0 for z in range(1, padding - keep1 + 1)],\n",
    "                \"dist\": range(keep1 + 1, padding + 1),\n",
    "            }\n",
    "        )\n",
    "    weight = pd.concat([weight1, weight2, weight3])\n",
    "    weight_dict = pd.Series(weight.weight.values, index=weight.dist).to_dict()\n",
    "    return weight_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "assured-upset",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load the table to convert the chromosome names to the USCS format\n",
    "CRCH38_ensembl2UCSC = pd.read_csv('/scratch/bacint/jdeleuw/KC_ATAC/analysis/bigBed/CRCH38_ensembl2UCSC.txt', sep ='\\t',header = 0, index_col=False)\n",
    "\n",
    "#make dictionary which can be used to convert ensembl to UCSC format \n",
    "ensembl_to_UCSC = {}\n",
    "UCSC_to_ensembl = {}\n",
    "for row in CRCH38_ensembl2UCSC.itertuples():\n",
    "    ensembl_to_UCSC[row[1]] = row[2]\n",
    "    UCSC_to_ensembl[row[2]] = row[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bored-advantage",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "spare-flesh",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import TFs of interest\n",
    "TFs = pd.read_csv('/scratch/bacint/jdeleuw/ananse/analysis/interesting_TFs.csv', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f4505a2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable_names = pd.read_csv('/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/Variable_names.csv', sep = ',', header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15ccff99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import binding files\n",
    "os.chdir('/scratch/bacint/jdeleuw/dev_ananse/binding_data')\n",
    "for index in variable_names.index:\n",
    "    path_list = [os.getcwd(), '/binding_', index, '/binding.tsv']\n",
    "    path = ''.join(path_list)\n",
    "    globals()[index] = dd.read_csv(path, sep='\\t', engine='python')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acknowledged-verse",
   "metadata": {},
   "source": [
    "# Filter TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff4edd91",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#filter TFs of interest from binding files\n",
    "binding_dfs = [globals()[i] for i in dir() if re.match('^day.', i)]\n",
    "binding_dfs.extend([globals()[i] for i in dir() if re.match('KC', i)])\n",
    "\n",
    "for name, binding_df in zip(variable_names.TF_filtered_binding_dfs, binding_dfs):\n",
    "    globals()[name] = binding_df[binding_df['factor'].isin(TFs[0].tolist())].compute()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "answering-springer",
   "metadata": {},
   "source": [
    "# Get all the enhancer within 100 kb window of TSS and calculate the distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ce704e7c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "TF_filtered_dfs = [globals()[i] for i in dir() if re.search('^filtered_Day.$', i) \n",
    "                   or re.search('^filtered_Day..$', i) \n",
    "                   or re.search('^filtered_KC$', i)]\n",
    "\n",
    "for df, day in zip(TF_filtered_dfs, variable_names.index):\n",
    "    enhancer_duplicates = df['enhancer'].str.split(':|-', n = 2, expand = True)\n",
    "    enhancer = enhancer_duplicates.drop_duplicates()\n",
    "    enhancer_ensembl = enhancer.replace({0 : UCSC_to_ensembl})\n",
    "    \n",
    "    path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/bed_files/', day, '.bed']\n",
    "    path = ''.join(path_list)\n",
    "    \n",
    "    enhancer_ensembl.to_csv(path, sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "existing-hardwood",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset the TSS of each gene from the GTF file\n",
    "genome_TSS_annotation('/scratch/bacint/jdeleuw/genomes/GRCh38.p13/GRCh38.p13.annotation.gtf', \n",
    "                      '/scratch/bacint/jdeleuw/genomes/GRCh38.p13/GRCh38.p13.small.annotation.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "missing-value",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make df with TSS and enhancers within 100 kb\n",
    "TF_filtered_dfs = [globals()[i] for i in dir() if re.search('^filtered_Day.$', i) \n",
    "                   or re.search('^filtered_Day..$', i) \n",
    "                   or re.search('^filtered_KC$', i)]\n",
    "\n",
    "for binding_df, day, output_name in zip(TF_filtered_dfs, variable_names.index, variable_names.distance_TSS_enhancer):\n",
    "    path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/bed_files/', day, '.bed']\n",
    "    path = ''.join(path_list)\n",
    "    \n",
    "    df = TSS_window_to_region('/scratch/bacint/jdeleuw/genomes/GRCh38.p13/GRCh38.p13.small.annotation.gtf',\n",
    "                                                   path,\n",
    "                                                   100000)\n",
    "    #filter TF of interest(influence_score >= 0.95) from the df\n",
    "    filtered_df = df[df['gene_name'].isin(TFs[0].tolist())]\n",
    "    \n",
    "    #merge df with TSS and linked enhancers, with the factor binding to the enhancer with the binding score\n",
    "    filtered_df = filtered_df.replace({'Chrom': ensembl_to_UCSC})\n",
    "    filtered_df['enhancer'] = filtered_df[['Chrom', 'ChromStart']].astype(str).agg(':'.join, axis=1)\n",
    "    filtered_df['enhancer'] = filtered_df[['enhancer', 'ChromEnd']].astype(str).agg('-'.join, axis=1)\n",
    "    \n",
    "    if day == 'day0':\n",
    "        merged_df_temp = filtered_df.merge(binding_df, how = 'inner', on = 'enhancer')\n",
    "        merged_df = merged_df_temp.rename(columns = {'binding' : 'binding_day0'})\n",
    "    else: \n",
    "        merged_df['binding_' + day] = filtered_df.merge(binding_df, how = 'inner', on = 'enhancer')['binding']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "juvenile-sheriff",
   "metadata": {},
   "source": [
    "# Normalize the binding score based on the distance\n",
    "# How are the TFs of each cluster targeted by all the TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "certified-location",
   "metadata": {},
   "outputs": [],
   "source": [
    "Cluster1 = ['NANOG', 'SOX2', 'SALL4', 'PRDM14', 'ZSCAN10', 'OTX2', 'ZIC3', 'ZFP42', 'MYCN', 'FOXH1']\n",
    "Cluster2 = ['MYC']\n",
    "Cluster3 = ['MEIS1', 'HAND1', 'MSX2', 'HOXB2', 'MEIS2', 'HEY1', 'TBX3', 'TFAP2B']\n",
    "Cluster4 = ['TP63', 'MAF', 'EHF', 'HES2']\n",
    "Cluster5 = ['EPAS1', 'CEBPD', 'KLF5', 'IRF6', 'GRHL1', 'FOSL2', 'SNAI2']\n",
    "Cluster6 = ['IRF9', 'IRX3', 'OSR1', 'HOXB1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "first-thomson",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add weight based on the distance from the binding site to the TSS\n",
    "Cluster_list = [globals()[i] for i in dir() if re.search('^Cluster.$', i)]\n",
    "\n",
    "weights = distance_weight(alpha=1e4, include_promoter=False)\n",
    "test = merged_df.copy()\n",
    "\n",
    "for Cluster, output_name in zip(Cluster_list, variable_names.targeted_cluster_hm):\n",
    "    hm_data = pd.DataFrame(index = TFs[0], columns = variable_names.index)\n",
    "    for day in variable_names.index:\n",
    "        test['weight'] = test['distance'].map(weights)\n",
    "        test['binding_weight_' + day] = test['binding_' + day] * test['weight']      \n",
    "\n",
    "        regulation_probability = pd.DataFrame(index = TFs[0], columns = TFs[0])\n",
    "        for target_gene, TF in itertools.product(Cluster, TFs[0]):\n",
    "            filtered_target = test[test['gene_name'].isin([target_gene])]\n",
    "            filtered_TF = filtered_target[filtered_target['factor'].isin([TF])]\n",
    "\n",
    "            regulation_probability.at[target_gene, TF] = filtered_TF['binding_weight_' + day].sum()\n",
    "        for TF in TFs[0]:\n",
    "            regulation_probability.replace(0, np.nan, inplace=True)\n",
    "            hm_data.at[TF, day] = regulation_probability.loc[:, TF].mean()\n",
    "        globals()[output_name] = hm_data\n",
    "        path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/', output_name, '.tsv']\n",
    "        path = ''.join(path_list)\n",
    "        hm_data.to_csv(path, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vocal-change",
   "metadata": {},
   "source": [
    "# from the old run everything before this point is identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1e806b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make df with TSS and enhancers within 100 kb\n",
    "TF_filtered_dfs = [globals()[i] for i in dir() if re.search('^filtered_Day.$', i) \n",
    "                   or re.search('^filtered_Day..$', i) \n",
    "                   or re.search('^filtered_KC$', i)]\n",
    "\n",
    "for binding_df, day, output_name in zip(TF_filtered_dfs, variable_names.index, variable_names.distance_TSS_enhancer):\n",
    "    path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/bed_files/', day, '.bed']\n",
    "    path = ''.join(path_list)\n",
    "    \n",
    "    df = TSS_window_to_region('/scratch/bacint/jdeleuw/genomes/GRCh38.p13/GRCh38.p13.small.annotation.gtf',\n",
    "                                                   path,\n",
    "                                                   100000)\n",
    "    #filter TF of interest(influence_score > 0.95) from te df\n",
    "    filtered_df = df[df['gene_name'].isin(TFs[0].tolist())]\n",
    "    \n",
    "    #merge df with TSS and linked enhancers, with the factor binding to the enhancer with the binding score\n",
    "    filtered_df = filtered_df.replace({'Chrom': ensembl_to_UCSC})\n",
    "    filtered_df['enhancer'] = filtered_df[['Chrom', 'ChromStart']].astype(str).agg(':'.join, axis=1)\n",
    "    filtered_df['enhancer'] = filtered_df[['enhancer', 'ChromEnd']].astype(str).agg('-'.join, axis=1)\n",
    "    \n",
    "    globals()[output_name] =  filtered_df.merge(binding_df, how = 'inner', on = 'enhancer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dad88a96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_TSS</th>\n",
       "      <th>start_TSS</th>\n",
       "      <th>end_TSS</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>ChromStart</th>\n",
       "      <th>ChromEnd</th>\n",
       "      <th>distance</th>\n",
       "      <th>enhancer</th>\n",
       "      <th>factor</th>\n",
       "      <th>binding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>6424669</td>\n",
       "      <td>6424670</td>\n",
       "      <td>HES2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6325544</td>\n",
       "      <td>6325745</td>\n",
       "      <td>99125</td>\n",
       "      <td>chr1:6325544-6325745</td>\n",
       "      <td>EPAS1</td>\n",
       "      <td>0.34548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>6424669</td>\n",
       "      <td>6424670</td>\n",
       "      <td>HES2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6325544</td>\n",
       "      <td>6325745</td>\n",
       "      <td>99125</td>\n",
       "      <td>chr1:6325544-6325745</td>\n",
       "      <td>SNAI2</td>\n",
       "      <td>0.18312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>6424669</td>\n",
       "      <td>6424670</td>\n",
       "      <td>HES2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6325544</td>\n",
       "      <td>6325745</td>\n",
       "      <td>99125</td>\n",
       "      <td>chr1:6325544-6325745</td>\n",
       "      <td>MSX2</td>\n",
       "      <td>0.20395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>6424669</td>\n",
       "      <td>6424670</td>\n",
       "      <td>HES2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6325544</td>\n",
       "      <td>6325745</td>\n",
       "      <td>99125</td>\n",
       "      <td>chr1:6325544-6325745</td>\n",
       "      <td>HOXB2</td>\n",
       "      <td>0.31641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>6424669</td>\n",
       "      <td>6424670</td>\n",
       "      <td>HES2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6325544</td>\n",
       "      <td>6325745</td>\n",
       "      <td>99125</td>\n",
       "      <td>chr1:6325544-6325745</td>\n",
       "      <td>NANOG</td>\n",
       "      <td>0.11331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29881</th>\n",
       "      <td>8</td>\n",
       "      <td>144475848</td>\n",
       "      <td>144475849</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>144575536</td>\n",
       "      <td>144575737</td>\n",
       "      <td>99688</td>\n",
       "      <td>chr8:144575536-144575737</td>\n",
       "      <td>GRHL1</td>\n",
       "      <td>0.18814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29882</th>\n",
       "      <td>8</td>\n",
       "      <td>144475848</td>\n",
       "      <td>144475849</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>144575536</td>\n",
       "      <td>144575737</td>\n",
       "      <td>99688</td>\n",
       "      <td>chr8:144575536-144575737</td>\n",
       "      <td>IRF9</td>\n",
       "      <td>0.43239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29883</th>\n",
       "      <td>8</td>\n",
       "      <td>144475848</td>\n",
       "      <td>144475849</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>144575536</td>\n",
       "      <td>144575737</td>\n",
       "      <td>99688</td>\n",
       "      <td>chr8:144575536-144575737</td>\n",
       "      <td>TBX3</td>\n",
       "      <td>0.45133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29884</th>\n",
       "      <td>8</td>\n",
       "      <td>144475848</td>\n",
       "      <td>144475849</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>144575536</td>\n",
       "      <td>144575737</td>\n",
       "      <td>99688</td>\n",
       "      <td>chr8:144575536-144575737</td>\n",
       "      <td>CEBPD</td>\n",
       "      <td>0.31083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29885</th>\n",
       "      <td>8</td>\n",
       "      <td>144475848</td>\n",
       "      <td>144475849</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>144575536</td>\n",
       "      <td>144575737</td>\n",
       "      <td>99688</td>\n",
       "      <td>chr8:144575536-144575737</td>\n",
       "      <td>FOXH1</td>\n",
       "      <td>0.71105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29886 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_TSS  start_TSS    end_TSS gene_name Chrom  ChromStart   ChromEnd  \\\n",
       "0           1    6424669    6424670      HES2  chr1     6325544    6325745   \n",
       "1           1    6424669    6424670      HES2  chr1     6325544    6325745   \n",
       "2           1    6424669    6424670      HES2  chr1     6325544    6325745   \n",
       "3           1    6424669    6424670      HES2  chr1     6325544    6325745   \n",
       "4           1    6424669    6424670      HES2  chr1     6325544    6325745   \n",
       "...       ...        ...        ...       ...   ...         ...        ...   \n",
       "29881       8  144475848  144475849     FOXH1  chr8   144575536  144575737   \n",
       "29882       8  144475848  144475849     FOXH1  chr8   144575536  144575737   \n",
       "29883       8  144475848  144475849     FOXH1  chr8   144575536  144575737   \n",
       "29884       8  144475848  144475849     FOXH1  chr8   144575536  144575737   \n",
       "29885       8  144475848  144475849     FOXH1  chr8   144575536  144575737   \n",
       "\n",
       "      distance                  enhancer factor  binding  \n",
       "0        99125      chr1:6325544-6325745  EPAS1  0.34548  \n",
       "1        99125      chr1:6325544-6325745  SNAI2  0.18312  \n",
       "2        99125      chr1:6325544-6325745   MSX2  0.20395  \n",
       "3        99125      chr1:6325544-6325745  HOXB2  0.31641  \n",
       "4        99125      chr1:6325544-6325745  NANOG  0.11331  \n",
       "...        ...                       ...    ...      ...  \n",
       "29881    99688  chr8:144575536-144575737  GRHL1  0.18814  \n",
       "29882    99688  chr8:144575536-144575737   IRF9  0.43239  \n",
       "29883    99688  chr8:144575536-144575737   TBX3  0.45133  \n",
       "29884    99688  chr8:144575536-144575737  CEBPD  0.31083  \n",
       "29885    99688  chr8:144575536-144575737  FOXH1  0.71105  \n",
       "\n",
       "[29886 rows x 11 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test = distance_Day0[distance_Day0['gene_name'].isin(['HES2'])]\n",
    "# test['enhancer'].unique()\n",
    "distance_Day0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e02e8b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "Cluster1 = ['NANOG', 'SOX2', 'SALL4', 'PRDM14', 'ZSCAN10', 'OTX2', 'ZIC3', 'ZFP42', 'MYCN', 'FOXH1']\n",
    "Cluster2 = ['MYC']\n",
    "Cluster3 = ['MEIS1', 'HAND1', 'MSX2', 'HOXB2', 'MEIS2', 'HEY1', 'TBX3', 'TFAP2B']\n",
    "Cluster4 = ['TP63', 'MAF', 'EHF', 'HES2']\n",
    "Cluster5 = ['EPAS1', 'CEBPD', 'KLF5', 'IRF6', 'GRHL1', 'FOSL2', 'SNAI2']\n",
    "Cluster6 = ['IRF9', 'IRX3', 'OSR1', 'HOXB1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "be9d2d91",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add weight based on the distance from the binding site to the TSS\n",
    "distance_dfs = [globals()[i] for i in dir() if re.search('^distance_Day.$', i)\n",
    "                or re.search('^distance_Day..$', i) \n",
    "                or re.search('^distance_KC$', i)]\n",
    "Cluster_list = [globals()[i] for i in dir() if re.search('^Cluster.$', i)]\n",
    "\n",
    "weights = distance_weight(alpha=1e4, include_promoter=False)\n",
    "\n",
    "for Cluster, output_name in zip(Cluster_list, variable_names.targeted_cluster_hm):\n",
    "    hm_data = pd.DataFrame(index = TFs[0], columns = variable_names.index)\n",
    "    for df, day in zip(distance_dfs, variable_names.index):\n",
    "        df['weight'] = df['distance'].map(weights)\n",
    "        df['binding_weight'] = df['binding'] * df['weight']      \n",
    "\n",
    "        regulation_probability = pd.DataFrame(index = TFs[0], columns = TFs[0])\n",
    "        for target_gene, TF in itertools.product(TFs[0], TFs[0]):\n",
    "            filtered_target = df[df['gene_name'].isin([target_gene])]\n",
    "            filtered_TF = filtered_target[filtered_target['factor'].isin([TF])]\n",
    "\n",
    "            regulation_probability.at[target_gene, TF] = filtered_TF['binding_weight'].sum()\n",
    "        for TF in TFs[0]:\n",
    "            regulation_probability.replace(0, np.nan, inplace=True)\n",
    "            hm_data.at[TF, day] = regulation_probability.loc[:, TF].median()\n",
    "    globals()[output_name] = hm_data\n",
    "    path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/', output_name, '2.tsv']\n",
    "    path = ''.join(path_list)\n",
    "    hm_data.to_csv(path, sep='\\t')\n",
    "# hm_data_all.to_csv('/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/avg_binding.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "061955b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Day</th>\n",
       "      <th>day0</th>\n",
       "      <th>day07</th>\n",
       "      <th>day14</th>\n",
       "      <th>day21</th>\n",
       "      <th>day43</th>\n",
       "      <th>day60</th>\n",
       "      <th>KC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CEBPD</th>\n",
       "      <td>1.288474</td>\n",
       "      <td>1.408854</td>\n",
       "      <td>1.335435</td>\n",
       "      <td>1.522592</td>\n",
       "      <td>1.370805</td>\n",
       "      <td>1.236190</td>\n",
       "      <td>1.357187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EHF</th>\n",
       "      <td>1.288540</td>\n",
       "      <td>1.531236</td>\n",
       "      <td>1.723425</td>\n",
       "      <td>1.569336</td>\n",
       "      <td>1.467538</td>\n",
       "      <td>1.417249</td>\n",
       "      <td>1.386874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EPAS1</th>\n",
       "      <td>1.876633</td>\n",
       "      <td>1.967900</td>\n",
       "      <td>1.948129</td>\n",
       "      <td>1.898744</td>\n",
       "      <td>1.960539</td>\n",
       "      <td>1.820026</td>\n",
       "      <td>1.911307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL2</th>\n",
       "      <td>1.678364</td>\n",
       "      <td>2.045805</td>\n",
       "      <td>1.810959</td>\n",
       "      <td>1.601668</td>\n",
       "      <td>1.772242</td>\n",
       "      <td>1.736251</td>\n",
       "      <td>1.664599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXH1</th>\n",
       "      <td>1.409630</td>\n",
       "      <td>1.720509</td>\n",
       "      <td>1.767127</td>\n",
       "      <td>1.620360</td>\n",
       "      <td>1.550774</td>\n",
       "      <td>1.486554</td>\n",
       "      <td>1.526097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRHL1</th>\n",
       "      <td>1.542192</td>\n",
       "      <td>1.790372</td>\n",
       "      <td>1.549551</td>\n",
       "      <td>1.741715</td>\n",
       "      <td>1.663205</td>\n",
       "      <td>1.582146</td>\n",
       "      <td>1.502585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HAND1</th>\n",
       "      <td>1.510161</td>\n",
       "      <td>1.570748</td>\n",
       "      <td>1.577544</td>\n",
       "      <td>1.676905</td>\n",
       "      <td>1.658619</td>\n",
       "      <td>1.462567</td>\n",
       "      <td>1.494010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES2</th>\n",
       "      <td>2.005505</td>\n",
       "      <td>2.021125</td>\n",
       "      <td>2.021099</td>\n",
       "      <td>2.051447</td>\n",
       "      <td>2.011806</td>\n",
       "      <td>2.004647</td>\n",
       "      <td>2.009485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEY1</th>\n",
       "      <td>1.429732</td>\n",
       "      <td>1.545142</td>\n",
       "      <td>1.676679</td>\n",
       "      <td>1.651011</td>\n",
       "      <td>1.465219</td>\n",
       "      <td>1.508449</td>\n",
       "      <td>1.447178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXB1</th>\n",
       "      <td>1.366174</td>\n",
       "      <td>1.605314</td>\n",
       "      <td>1.683612</td>\n",
       "      <td>1.744091</td>\n",
       "      <td>1.560055</td>\n",
       "      <td>1.430650</td>\n",
       "      <td>1.437270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXB2</th>\n",
       "      <td>2.315030</td>\n",
       "      <td>2.315008</td>\n",
       "      <td>2.315023</td>\n",
       "      <td>2.314839</td>\n",
       "      <td>2.314963</td>\n",
       "      <td>2.315042</td>\n",
       "      <td>2.314988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRF6</th>\n",
       "      <td>2.251276</td>\n",
       "      <td>2.261883</td>\n",
       "      <td>2.255421</td>\n",
       "      <td>2.336494</td>\n",
       "      <td>2.285739</td>\n",
       "      <td>2.247553</td>\n",
       "      <td>2.277238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRF9</th>\n",
       "      <td>2.068792</td>\n",
       "      <td>2.156435</td>\n",
       "      <td>2.160370</td>\n",
       "      <td>2.216028</td>\n",
       "      <td>2.074322</td>\n",
       "      <td>2.104400</td>\n",
       "      <td>2.097667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRX3</th>\n",
       "      <td>1.475064</td>\n",
       "      <td>1.517951</td>\n",
       "      <td>1.607159</td>\n",
       "      <td>1.520798</td>\n",
       "      <td>1.655397</td>\n",
       "      <td>1.437390</td>\n",
       "      <td>1.458734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF5</th>\n",
       "      <td>1.572000</td>\n",
       "      <td>1.669735</td>\n",
       "      <td>1.619314</td>\n",
       "      <td>1.741736</td>\n",
       "      <td>1.659450</td>\n",
       "      <td>1.585408</td>\n",
       "      <td>1.587476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAF</th>\n",
       "      <td>1.313203</td>\n",
       "      <td>1.536844</td>\n",
       "      <td>1.498194</td>\n",
       "      <td>1.414446</td>\n",
       "      <td>1.457743</td>\n",
       "      <td>1.426675</td>\n",
       "      <td>1.281279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEIS1</th>\n",
       "      <td>1.771333</td>\n",
       "      <td>1.966896</td>\n",
       "      <td>1.904402</td>\n",
       "      <td>2.071686</td>\n",
       "      <td>1.893693</td>\n",
       "      <td>1.822869</td>\n",
       "      <td>1.832032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEIS2</th>\n",
       "      <td>1.861103</td>\n",
       "      <td>1.842828</td>\n",
       "      <td>1.880619</td>\n",
       "      <td>1.888029</td>\n",
       "      <td>1.920700</td>\n",
       "      <td>1.822550</td>\n",
       "      <td>1.718274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSX2</th>\n",
       "      <td>1.378441</td>\n",
       "      <td>1.589222</td>\n",
       "      <td>1.675112</td>\n",
       "      <td>1.748040</td>\n",
       "      <td>1.552238</td>\n",
       "      <td>1.388005</td>\n",
       "      <td>1.425325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYC</th>\n",
       "      <td>1.459077</td>\n",
       "      <td>1.681254</td>\n",
       "      <td>1.658942</td>\n",
       "      <td>1.614066</td>\n",
       "      <td>1.532889</td>\n",
       "      <td>1.518196</td>\n",
       "      <td>1.425092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCN</th>\n",
       "      <td>1.683788</td>\n",
       "      <td>1.710703</td>\n",
       "      <td>1.705831</td>\n",
       "      <td>1.801780</td>\n",
       "      <td>1.716649</td>\n",
       "      <td>1.647079</td>\n",
       "      <td>1.705900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NANOG</th>\n",
       "      <td>1.453950</td>\n",
       "      <td>1.450667</td>\n",
       "      <td>1.481561</td>\n",
       "      <td>1.405860</td>\n",
       "      <td>1.429549</td>\n",
       "      <td>1.539945</td>\n",
       "      <td>1.414304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OSR1</th>\n",
       "      <td>1.498962</td>\n",
       "      <td>1.658516</td>\n",
       "      <td>1.613629</td>\n",
       "      <td>1.633649</td>\n",
       "      <td>1.624949</td>\n",
       "      <td>1.564528</td>\n",
       "      <td>1.685133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTX2</th>\n",
       "      <td>1.366102</td>\n",
       "      <td>1.612651</td>\n",
       "      <td>1.701848</td>\n",
       "      <td>1.698917</td>\n",
       "      <td>1.531763</td>\n",
       "      <td>1.372234</td>\n",
       "      <td>1.475191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRDM14</th>\n",
       "      <td>1.363541</td>\n",
       "      <td>1.532938</td>\n",
       "      <td>1.575043</td>\n",
       "      <td>1.350468</td>\n",
       "      <td>1.399436</td>\n",
       "      <td>1.300363</td>\n",
       "      <td>1.356740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SALL4</th>\n",
       "      <td>1.658163</td>\n",
       "      <td>1.743854</td>\n",
       "      <td>1.666499</td>\n",
       "      <td>1.830686</td>\n",
       "      <td>1.772623</td>\n",
       "      <td>1.672869</td>\n",
       "      <td>1.728571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNAI2</th>\n",
       "      <td>1.261624</td>\n",
       "      <td>1.441864</td>\n",
       "      <td>1.478664</td>\n",
       "      <td>1.478388</td>\n",
       "      <td>1.470870</td>\n",
       "      <td>1.297594</td>\n",
       "      <td>1.394248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SOX2</th>\n",
       "      <td>1.351240</td>\n",
       "      <td>1.457366</td>\n",
       "      <td>1.611395</td>\n",
       "      <td>1.566150</td>\n",
       "      <td>1.544220</td>\n",
       "      <td>1.434724</td>\n",
       "      <td>1.440143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBX3</th>\n",
       "      <td>2.133333</td>\n",
       "      <td>2.127486</td>\n",
       "      <td>2.187454</td>\n",
       "      <td>2.238283</td>\n",
       "      <td>2.161902</td>\n",
       "      <td>2.104733</td>\n",
       "      <td>2.149281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFAP2B</th>\n",
       "      <td>1.646548</td>\n",
       "      <td>1.874279</td>\n",
       "      <td>1.889701</td>\n",
       "      <td>1.931949</td>\n",
       "      <td>1.822246</td>\n",
       "      <td>1.641136</td>\n",
       "      <td>1.684865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP63</th>\n",
       "      <td>0.757821</td>\n",
       "      <td>0.967079</td>\n",
       "      <td>0.908421</td>\n",
       "      <td>0.815227</td>\n",
       "      <td>0.820936</td>\n",
       "      <td>0.688594</td>\n",
       "      <td>0.615954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFP42</th>\n",
       "      <td>0.800631</td>\n",
       "      <td>0.874786</td>\n",
       "      <td>0.943423</td>\n",
       "      <td>0.951780</td>\n",
       "      <td>0.992837</td>\n",
       "      <td>0.954656</td>\n",
       "      <td>0.969090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>1.725126</td>\n",
       "      <td>1.827106</td>\n",
       "      <td>1.782717</td>\n",
       "      <td>1.862219</td>\n",
       "      <td>1.758467</td>\n",
       "      <td>1.758611</td>\n",
       "      <td>1.754567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN10</th>\n",
       "      <td>1.479936</td>\n",
       "      <td>1.695013</td>\n",
       "      <td>1.606829</td>\n",
       "      <td>1.481060</td>\n",
       "      <td>1.534253</td>\n",
       "      <td>1.504594</td>\n",
       "      <td>1.362689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Day          day0     day07     day14     day21     day43     day60        KC\n",
       "0                                                                            \n",
       "CEBPD    1.288474  1.408854  1.335435  1.522592  1.370805  1.236190  1.357187\n",
       "EHF      1.288540  1.531236  1.723425  1.569336  1.467538  1.417249  1.386874\n",
       "EPAS1    1.876633  1.967900  1.948129  1.898744  1.960539  1.820026  1.911307\n",
       "FOSL2    1.678364  2.045805  1.810959  1.601668  1.772242  1.736251  1.664599\n",
       "FOXH1    1.409630  1.720509  1.767127  1.620360  1.550774  1.486554  1.526097\n",
       "GRHL1    1.542192  1.790372  1.549551  1.741715  1.663205  1.582146  1.502585\n",
       "HAND1    1.510161  1.570748  1.577544  1.676905  1.658619  1.462567  1.494010\n",
       "HES2     2.005505  2.021125  2.021099  2.051447  2.011806  2.004647  2.009485\n",
       "HEY1     1.429732  1.545142  1.676679  1.651011  1.465219  1.508449  1.447178\n",
       "HOXB1    1.366174  1.605314  1.683612  1.744091  1.560055  1.430650  1.437270\n",
       "HOXB2    2.315030  2.315008  2.315023  2.314839  2.314963  2.315042  2.314988\n",
       "IRF6     2.251276  2.261883  2.255421  2.336494  2.285739  2.247553  2.277238\n",
       "IRF9     2.068792  2.156435  2.160370  2.216028  2.074322  2.104400  2.097667\n",
       "IRX3     1.475064  1.517951  1.607159  1.520798  1.655397  1.437390  1.458734\n",
       "KLF5     1.572000  1.669735  1.619314  1.741736  1.659450  1.585408  1.587476\n",
       "MAF      1.313203  1.536844  1.498194  1.414446  1.457743  1.426675  1.281279\n",
       "MEIS1    1.771333  1.966896  1.904402  2.071686  1.893693  1.822869  1.832032\n",
       "MEIS2    1.861103  1.842828  1.880619  1.888029  1.920700  1.822550  1.718274\n",
       "MSX2     1.378441  1.589222  1.675112  1.748040  1.552238  1.388005  1.425325\n",
       "MYC      1.459077  1.681254  1.658942  1.614066  1.532889  1.518196  1.425092\n",
       "MYCN     1.683788  1.710703  1.705831  1.801780  1.716649  1.647079  1.705900\n",
       "NANOG    1.453950  1.450667  1.481561  1.405860  1.429549  1.539945  1.414304\n",
       "OSR1     1.498962  1.658516  1.613629  1.633649  1.624949  1.564528  1.685133\n",
       "OTX2     1.366102  1.612651  1.701848  1.698917  1.531763  1.372234  1.475191\n",
       "PRDM14   1.363541  1.532938  1.575043  1.350468  1.399436  1.300363  1.356740\n",
       "SALL4    1.658163  1.743854  1.666499  1.830686  1.772623  1.672869  1.728571\n",
       "SNAI2    1.261624  1.441864  1.478664  1.478388  1.470870  1.297594  1.394248\n",
       "SOX2     1.351240  1.457366  1.611395  1.566150  1.544220  1.434724  1.440143\n",
       "TBX3     2.133333  2.127486  2.187454  2.238283  2.161902  2.104733  2.149281\n",
       "TFAP2B   1.646548  1.874279  1.889701  1.931949  1.822246  1.641136  1.684865\n",
       "TP63     0.757821  0.967079  0.908421  0.815227  0.820936  0.688594  0.615954\n",
       "ZFP42    0.800631  0.874786  0.943423  0.951780  0.992837  0.954656  0.969090\n",
       "ZIC3     1.725126  1.827106  1.782717  1.862219  1.758467  1.758611  1.754567\n",
       "ZSCAN10  1.479936  1.695013  1.606829  1.481060  1.534253  1.504594  1.362689"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hm_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "593c626c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add weight based on the distance from the binding site to the TSS\n",
    "distance_dfs = [globals()[i] for i in dir() if re.search('^distance_Day.$', i)\n",
    "                or re.search('^distance_Day..$', i) \n",
    "                or re.search('^distance_KC$', i)]\n",
    "Cluster_list = [globals()[i] for i in dir() if re.search('^Cluster.$', i)]\n",
    "\n",
    "weights = distance_weight(alpha=1e4, include_promoter=False)\n",
    "\n",
    "for Cluster, output_name in zip(Cluster_list, variable_names.targeting_cluster_hm):\n",
    "    hm_data = pd.DataFrame(index = TFs[0], columns = variable_names.index)\n",
    "    for df, day in zip(distance_dfs, variable_names.index):\n",
    "        df['weight'] = df['distance'].map(weights)\n",
    "        df['binding_weight'] = df['binding'] * df['weight']      \n",
    "        \n",
    "        regulation_probability = pd.DataFrame(index = TFs[0], columns = TFs[0])\n",
    "        for target_gene, TF in itertools.product(Cluster, TFs[0]):\n",
    "            filtered_target = df[df['gene_name'].isin([target_gene])]\n",
    "            filtered_TF = filtered_target[filtered_target['factor'].isin([TF])]\n",
    "\n",
    "            regulation_probability.at[target_gene, TF] = filtered_TF['binding_weight'].sum()\n",
    "        for TF in TFs[0]:\n",
    "            regulation_probability.replace(0, np.nan, inplace=True)\n",
    "            hm_data.at[TF, day] = regulation_probability.loc[:, TF].mean()\n",
    "    globals()[output_name] = hm_data\n",
    "    path_list = ['/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/', output_name, '2.tsv']\n",
    "    path = ''.join(path_list)\n",
    "    hm_data.to_csv(path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "af8167c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Day</th>\n",
       "      <th>day0</th>\n",
       "      <th>day07</th>\n",
       "      <th>day14</th>\n",
       "      <th>day21</th>\n",
       "      <th>day43</th>\n",
       "      <th>day60</th>\n",
       "      <th>KC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CEBPD</th>\n",
       "      <td>1.789992</td>\n",
       "      <td>2.356418</td>\n",
       "      <td>1.934651</td>\n",
       "      <td>1.714610</td>\n",
       "      <td>1.519043</td>\n",
       "      <td>1.538332</td>\n",
       "      <td>1.455296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EHF</th>\n",
       "      <td>1.892087</td>\n",
       "      <td>2.556373</td>\n",
       "      <td>2.110531</td>\n",
       "      <td>1.840660</td>\n",
       "      <td>1.544962</td>\n",
       "      <td>1.573766</td>\n",
       "      <td>1.518893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EPAS1</th>\n",
       "      <td>2.098745</td>\n",
       "      <td>2.592621</td>\n",
       "      <td>2.252064</td>\n",
       "      <td>2.050217</td>\n",
       "      <td>1.860546</td>\n",
       "      <td>1.897103</td>\n",
       "      <td>1.826770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOSL2</th>\n",
       "      <td>1.690914</td>\n",
       "      <td>2.032707</td>\n",
       "      <td>1.759675</td>\n",
       "      <td>1.596311</td>\n",
       "      <td>1.501649</td>\n",
       "      <td>1.550880</td>\n",
       "      <td>1.468265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXH1</th>\n",
       "      <td>1.749085</td>\n",
       "      <td>2.259563</td>\n",
       "      <td>1.859677</td>\n",
       "      <td>1.666911</td>\n",
       "      <td>1.526133</td>\n",
       "      <td>1.560329</td>\n",
       "      <td>1.468322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRHL1</th>\n",
       "      <td>1.900212</td>\n",
       "      <td>2.424015</td>\n",
       "      <td>2.048217</td>\n",
       "      <td>1.819320</td>\n",
       "      <td>1.611624</td>\n",
       "      <td>1.681833</td>\n",
       "      <td>1.584648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HAND1</th>\n",
       "      <td>1.765074</td>\n",
       "      <td>2.252119</td>\n",
       "      <td>1.915584</td>\n",
       "      <td>1.689119</td>\n",
       "      <td>1.522303</td>\n",
       "      <td>1.580482</td>\n",
       "      <td>1.496332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES2</th>\n",
       "      <td>1.940485</td>\n",
       "      <td>2.180157</td>\n",
       "      <td>2.024199</td>\n",
       "      <td>1.925941</td>\n",
       "      <td>1.842738</td>\n",
       "      <td>1.862760</td>\n",
       "      <td>1.803689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEY1</th>\n",
       "      <td>1.795266</td>\n",
       "      <td>2.319263</td>\n",
       "      <td>1.932877</td>\n",
       "      <td>1.751036</td>\n",
       "      <td>1.582067</td>\n",
       "      <td>1.618294</td>\n",
       "      <td>1.529095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXB1</th>\n",
       "      <td>1.893505</td>\n",
       "      <td>2.530493</td>\n",
       "      <td>2.115316</td>\n",
       "      <td>1.849169</td>\n",
       "      <td>1.575651</td>\n",
       "      <td>1.605443</td>\n",
       "      <td>1.535002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXB2</th>\n",
       "      <td>2.018086</td>\n",
       "      <td>2.017942</td>\n",
       "      <td>2.018006</td>\n",
       "      <td>2.018078</td>\n",
       "      <td>2.018140</td>\n",
       "      <td>2.018130</td>\n",
       "      <td>2.018168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRF6</th>\n",
       "      <td>1.996139</td>\n",
       "      <td>2.062017</td>\n",
       "      <td>2.029836</td>\n",
       "      <td>1.999858</td>\n",
       "      <td>1.970909</td>\n",
       "      <td>1.974807</td>\n",
       "      <td>1.957997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRF9</th>\n",
       "      <td>2.093189</td>\n",
       "      <td>2.405310</td>\n",
       "      <td>2.186493</td>\n",
       "      <td>2.059951</td>\n",
       "      <td>1.954853</td>\n",
       "      <td>1.983713</td>\n",
       "      <td>1.915991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IRX3</th>\n",
       "      <td>1.745993</td>\n",
       "      <td>2.279044</td>\n",
       "      <td>1.909003</td>\n",
       "      <td>1.689716</td>\n",
       "      <td>1.517494</td>\n",
       "      <td>1.567248</td>\n",
       "      <td>1.470443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF5</th>\n",
       "      <td>1.705144</td>\n",
       "      <td>2.037836</td>\n",
       "      <td>1.805058</td>\n",
       "      <td>1.673741</td>\n",
       "      <td>1.570544</td>\n",
       "      <td>1.604535</td>\n",
       "      <td>1.524968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAF</th>\n",
       "      <td>1.363628</td>\n",
       "      <td>1.540927</td>\n",
       "      <td>1.392809</td>\n",
       "      <td>1.294661</td>\n",
       "      <td>1.225080</td>\n",
       "      <td>1.240179</td>\n",
       "      <td>1.158518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEIS1</th>\n",
       "      <td>2.064851</td>\n",
       "      <td>2.459822</td>\n",
       "      <td>2.195239</td>\n",
       "      <td>2.035382</td>\n",
       "      <td>1.905830</td>\n",
       "      <td>1.949572</td>\n",
       "      <td>1.862195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEIS2</th>\n",
       "      <td>2.034529</td>\n",
       "      <td>2.430868</td>\n",
       "      <td>2.161112</td>\n",
       "      <td>2.005310</td>\n",
       "      <td>1.875476</td>\n",
       "      <td>1.918883</td>\n",
       "      <td>1.831094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSX2</th>\n",
       "      <td>1.898764</td>\n",
       "      <td>2.536856</td>\n",
       "      <td>2.120030</td>\n",
       "      <td>1.855903</td>\n",
       "      <td>1.582284</td>\n",
       "      <td>1.612378</td>\n",
       "      <td>1.542795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYC</th>\n",
       "      <td>1.798816</td>\n",
       "      <td>2.310719</td>\n",
       "      <td>1.934065</td>\n",
       "      <td>1.729879</td>\n",
       "      <td>1.554259</td>\n",
       "      <td>1.605711</td>\n",
       "      <td>1.519385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCN</th>\n",
       "      <td>1.772516</td>\n",
       "      <td>2.087958</td>\n",
       "      <td>1.867540</td>\n",
       "      <td>1.742367</td>\n",
       "      <td>1.642715</td>\n",
       "      <td>1.671687</td>\n",
       "      <td>1.596009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NANOG</th>\n",
       "      <td>1.736168</td>\n",
       "      <td>2.271698</td>\n",
       "      <td>1.870037</td>\n",
       "      <td>1.660567</td>\n",
       "      <td>1.498415</td>\n",
       "      <td>1.562729</td>\n",
       "      <td>1.480352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OSR1</th>\n",
       "      <td>1.848492</td>\n",
       "      <td>2.277148</td>\n",
       "      <td>1.929610</td>\n",
       "      <td>1.745787</td>\n",
       "      <td>1.587033</td>\n",
       "      <td>1.631214</td>\n",
       "      <td>1.556968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTX2</th>\n",
       "      <td>1.866993</td>\n",
       "      <td>2.507029</td>\n",
       "      <td>2.080452</td>\n",
       "      <td>1.817568</td>\n",
       "      <td>1.546736</td>\n",
       "      <td>1.586102</td>\n",
       "      <td>1.511099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRDM14</th>\n",
       "      <td>1.586322</td>\n",
       "      <td>2.025356</td>\n",
       "      <td>1.716885</td>\n",
       "      <td>1.509428</td>\n",
       "      <td>1.353753</td>\n",
       "      <td>1.435852</td>\n",
       "      <td>1.320649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SALL4</th>\n",
       "      <td>1.765094</td>\n",
       "      <td>2.074732</td>\n",
       "      <td>1.858627</td>\n",
       "      <td>1.740633</td>\n",
       "      <td>1.644437</td>\n",
       "      <td>1.673532</td>\n",
       "      <td>1.603657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNAI2</th>\n",
       "      <td>1.673248</td>\n",
       "      <td>2.186578</td>\n",
       "      <td>1.818163</td>\n",
       "      <td>1.593048</td>\n",
       "      <td>1.418054</td>\n",
       "      <td>1.478435</td>\n",
       "      <td>1.368934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SOX2</th>\n",
       "      <td>1.657838</td>\n",
       "      <td>2.283203</td>\n",
       "      <td>1.850938</td>\n",
       "      <td>1.614576</td>\n",
       "      <td>1.416773</td>\n",
       "      <td>1.483181</td>\n",
       "      <td>1.383680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBX3</th>\n",
       "      <td>1.948361</td>\n",
       "      <td>2.047741</td>\n",
       "      <td>1.996865</td>\n",
       "      <td>1.952654</td>\n",
       "      <td>1.910608</td>\n",
       "      <td>1.916913</td>\n",
       "      <td>1.891628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFAP2B</th>\n",
       "      <td>2.035955</td>\n",
       "      <td>2.515570</td>\n",
       "      <td>2.189511</td>\n",
       "      <td>1.986699</td>\n",
       "      <td>1.788393</td>\n",
       "      <td>1.816461</td>\n",
       "      <td>1.738179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP63</th>\n",
       "      <td>1.106845</td>\n",
       "      <td>1.544512</td>\n",
       "      <td>1.168961</td>\n",
       "      <td>1.003948</td>\n",
       "      <td>0.908787</td>\n",
       "      <td>1.001720</td>\n",
       "      <td>0.862346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFP42</th>\n",
       "      <td>1.732553</td>\n",
       "      <td>2.051294</td>\n",
       "      <td>1.753803</td>\n",
       "      <td>1.671454</td>\n",
       "      <td>1.550098</td>\n",
       "      <td>1.553426</td>\n",
       "      <td>1.487296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>1.806474</td>\n",
       "      <td>2.104959</td>\n",
       "      <td>1.903730</td>\n",
       "      <td>1.784761</td>\n",
       "      <td>1.688209</td>\n",
       "      <td>1.716064</td>\n",
       "      <td>1.644648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN10</th>\n",
       "      <td>1.632798</td>\n",
       "      <td>2.146336</td>\n",
       "      <td>1.781871</td>\n",
       "      <td>1.575197</td>\n",
       "      <td>1.412820</td>\n",
       "      <td>1.475375</td>\n",
       "      <td>1.382986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Day          day0     day07     day14     day21     day43     day60        KC\n",
       "0                                                                            \n",
       "CEBPD    1.789992  2.356418  1.934651  1.714610  1.519043  1.538332  1.455296\n",
       "EHF      1.892087  2.556373  2.110531  1.840660  1.544962  1.573766  1.518893\n",
       "EPAS1    2.098745  2.592621  2.252064  2.050217  1.860546  1.897103  1.826770\n",
       "FOSL2    1.690914  2.032707  1.759675  1.596311  1.501649  1.550880  1.468265\n",
       "FOXH1    1.749085  2.259563  1.859677  1.666911  1.526133  1.560329  1.468322\n",
       "GRHL1    1.900212  2.424015  2.048217  1.819320  1.611624  1.681833  1.584648\n",
       "HAND1    1.765074  2.252119  1.915584  1.689119  1.522303  1.580482  1.496332\n",
       "HES2     1.940485  2.180157  2.024199  1.925941  1.842738  1.862760  1.803689\n",
       "HEY1     1.795266  2.319263  1.932877  1.751036  1.582067  1.618294  1.529095\n",
       "HOXB1    1.893505  2.530493  2.115316  1.849169  1.575651  1.605443  1.535002\n",
       "HOXB2    2.018086  2.017942  2.018006  2.018078  2.018140  2.018130  2.018168\n",
       "IRF6     1.996139  2.062017  2.029836  1.999858  1.970909  1.974807  1.957997\n",
       "IRF9     2.093189  2.405310  2.186493  2.059951  1.954853  1.983713  1.915991\n",
       "IRX3     1.745993  2.279044  1.909003  1.689716  1.517494  1.567248  1.470443\n",
       "KLF5     1.705144  2.037836  1.805058  1.673741  1.570544  1.604535  1.524968\n",
       "MAF      1.363628  1.540927  1.392809  1.294661  1.225080  1.240179  1.158518\n",
       "MEIS1    2.064851  2.459822  2.195239  2.035382  1.905830  1.949572  1.862195\n",
       "MEIS2    2.034529  2.430868  2.161112  2.005310  1.875476  1.918883  1.831094\n",
       "MSX2     1.898764  2.536856  2.120030  1.855903  1.582284  1.612378  1.542795\n",
       "MYC      1.798816  2.310719  1.934065  1.729879  1.554259  1.605711  1.519385\n",
       "MYCN     1.772516  2.087958  1.867540  1.742367  1.642715  1.671687  1.596009\n",
       "NANOG    1.736168  2.271698  1.870037  1.660567  1.498415  1.562729  1.480352\n",
       "OSR1     1.848492  2.277148  1.929610  1.745787  1.587033  1.631214  1.556968\n",
       "OTX2     1.866993  2.507029  2.080452  1.817568  1.546736  1.586102  1.511099\n",
       "PRDM14   1.586322  2.025356  1.716885  1.509428  1.353753  1.435852  1.320649\n",
       "SALL4    1.765094  2.074732  1.858627  1.740633  1.644437  1.673532  1.603657\n",
       "SNAI2    1.673248  2.186578  1.818163  1.593048  1.418054  1.478435  1.368934\n",
       "SOX2     1.657838  2.283203  1.850938  1.614576  1.416773  1.483181  1.383680\n",
       "TBX3     1.948361  2.047741  1.996865  1.952654  1.910608  1.916913  1.891628\n",
       "TFAP2B   2.035955  2.515570  2.189511  1.986699  1.788393  1.816461  1.738179\n",
       "TP63     1.106845  1.544512  1.168961  1.003948  0.908787  1.001720  0.862346\n",
       "ZFP42    1.732553  2.051294  1.753803  1.671454  1.550098  1.553426  1.487296\n",
       "ZIC3     1.806474  2.104959  1.903730  1.784761  1.688209  1.716064  1.644648\n",
       "ZSCAN10  1.632798  2.146336  1.781871  1.575197  1.412820  1.475375  1.382986"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hm_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "wrong-prince",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'dist'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-bf6ebcbe6013>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# import from Ananse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mweight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdistance_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malpha\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minclude_promoter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlineplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"dist\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"weight\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"/scratch/bacint/jdeleuw/ananse/binding_data/binding_analysis/weight_distance_enhancer_window.pdf\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0menh_weight_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'dist'"
     ]
    }
   ],
   "source": [
    "sns.set(rc={\"figure.figsize\": (15, 3)})\n",
    "# import from Ananse\n",
    "weight = distance_weight(alpha=1e4, include_promoter=False)\n",
    "sns.lineplot(x=weight[\"dist\"], y=weight[\"weight\"])\n",
    "plt.savefig(f\"/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/weight_distance_enhancer_window.pdf\")\n",
    "enh_weight_dict = pd.Series(weight.weight.values, index=weight.dist).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "4cfc0fd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add weight based on the distance from the binding site to the TSS normalize it by substracting row means\n",
    "distance_dfs = [globals()[i] for i in dir() if re.search('^distance_Day.$', i)\n",
    "                or re.search('^distance_Day..$', i) \n",
    "                or re.search('^distance_KC$', i)]\n",
    "Cluster_list = [globals()[i] for i in dir() if re.search('^Cluster.$', i)]\n",
    "\n",
    "weights = distance_weight(alpha=1e4, include_promoter=False)\n",
    "\n",
    "# for Cluster, output_name in zip(Cluster_list, variable_names.targeted_cluster_hm):\n",
    "hm_data_row_mean = pd.DataFrame(index = TFs[0], columns = variable_names.index)\n",
    "for df, day in zip(distance_dfs, variable_names.index):\n",
    "    df['weight'] = df['distance'].map(weights)\n",
    "    df['binding_weight'] = df['binding'] * df['weight']      \n",
    "\n",
    "    regulation_probability = pd.DataFrame(index = TFs[0], columns = TFs[0])\n",
    "    for target_gene, TF in itertools.product(TFs[0], TFs[0]):\n",
    "        filtered_target = df[df['gene_name'].isin([target_gene])]\n",
    "        filtered_TF = filtered_target[filtered_target['factor'].isin([TF])]\n",
    "\n",
    "        regulation_probability.at[target_gene, TF] = filtered_TF['binding_weight'].sum()\n",
    "    for TF in TFs[0]:\n",
    "        regulation_probability.replace(0, np.nan, inplace=True)\n",
    "        #substract row_means from the values of each row\n",
    "        row_means = regulation_probability.mean(axis=1)\n",
    "        norm_regulation_probability = regulation_probability.loc[TF, :] - regulation_probability.mean(axis=1)[TF]\n",
    "                \n",
    "        hm_data_row_mean.at[TF, day] = norm_regulation_probability.mean()\n",
    "hm_data_row_mean.to_csv('/scratch/bacint/jdeleuw/dev_ananse/binding_data/binding_analysis/avg_binding_row_mean.tsv', sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
